velox/functions/prestosql/tests/StringFunctionsTest.cpp

```diff
@@ -1862,3 +1862,41 @@ TEST_F(StringFunctionsTest, varbinaryLength) {
   auto result = evaluate("length(c0)", makeRowVector({vector}));
   test::assertEqualVectors(expected, result);
 }
+
+TEST_F(StringFunctionsTest, normalize) {
+  const auto normalizeWithoutForm = [&](std::optional<std::string> string) {
+    return evaluateOnce<std::string>("normalize(c0)", string);
+  };
+
+  const auto normalizeWithForm = [&](std::optional<std::string> string,
+                                     const std::string& form) {
+    return evaluateOnce<std::string>(
+        fmt::format("normalize(c0, '{}')", form), string);
+  };
+
+  EXPECT_EQ(normalizeWithoutForm(std::nullopt), std::nullopt);
+  EXPECT_EQ(normalizeWithoutForm(""), "");
+  EXPECT_EQ(normalizeWithoutForm("sch\u00f6n"), "sch\u00f6n");
+  EXPECT_EQ(normalizeWithForm(std::nullopt, "NFD"), std::nullopt);
+  EXPECT_EQ(normalizeWithForm("", "NFKC"), "");
+  EXPECT_EQ(normalizeWithForm("sch\u00f6n", "NFD"), "scho\u0308n");
+  EXPECT_EQ(normalizeWithForm("sch\u00f6n", "NFC"), "sch\u00f6n");
+  EXPECT_EQ(normalizeWithForm("sch\u00f6n", "NFKD"), "scho\u0308n");
+  EXPECT_EQ(normalizeWithForm("sch\u00f6n", "NFKC"), "sch\u00f6n");
+  EXPECT_EQ(
+      normalizeWithForm("Hello world from Velox!!", "NFKC"),
+      "Hello world from Velox!!");
+  EXPECT_EQ(
+      normalizeWithForm(
```

Comment:
I'm wondering if these tests can be written in a way to self-check e.g. in a pattern like

normalize(normalize('value', 'some option'). 'other option') -> value

That would make it easier to get more confidence in them.

Reply:
The way I am writing tests now is:
-  Using a non-normalized string to get a normalized form of the string first. In this case it is `NFKC`.
-  Using the normalized string, writing test cases following the pattern `normalize(normalize('value', 'some option'). 'other option') -> value`. The value returned should be the normalized string.

```
 EXPECT_EQ(
      normalizeWithForm(
          "\uff8a\uff9d\uff76\uff78\uff76\uff85\uff8a\uff9d\uff76\uff78\uff76\uff85\uff8a\uff9d\uff76\uff78\uff76\uff85\uff8a\uff9d\uff76\uff78\uff76\uff85",
          "NFKC"),
      "\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca");

  EXPECT_EQ(
      normalizeWithForm(
          normalizeWithForm(
              "\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca\u30cf\u30f3\u30ab\u30af\u30ab\u30ca",
              "NFD"),
          "NFKC"),
```

Reply:
Thanks @pdabre12. That's definitely a more verifiable test. Can you add value variables for the literal values in the tests. They are hard to read and re-using a variable would make the test intent more clear.

Reply:
@aditi-pandit  Thanks for the review.
Added value variables for the literal values in the tests, PTAL.